lib/Dialect/TritonGPU/Transforms/OptimizeDotOperands.cpp

```diff
@@ -15,6 +17,120 @@ namespace gpu {
 
 namespace {
 
+// Helpers
+
+// Returns whether we can hoist DotOp Encoding through `op`.
+// Roughly, whether op is elementwise and thus threads don't need
+// to exchange elements. But some ops are not currently supported even though
+// they meet that criterion.
+bool canHoistDotOpEncV2(Operation* op, DotOperandEncodingAttr& dotOpEnc) {
+  // Only consider custom conversions or arith ops.
+  // TODO(jlebar): Is this too restrictive?
+  if (!isa<FpToFpOp, BitcastOp>(op) && !isPureUnaryInlineAsm(op) &&
+      !isa<arith::ArithDialect>(op->getDialect()))
+    return false;
+
+  // Quick handling to fix loading issues when computing the original
+  // bitwidth is unable to realize that there is a mixed-precision dot
+  // (hence kWidth = 1) but wants to hoist through the type conversion.
+  if (isa<arith::ExtFOp>(op) && dotOpEnc.getKWidth() == 1)
+      return false;
+
+  // Currently, these instructions are not supported during lowering of
+  // shared -> dot_operand layout. Not all types and type conversions are
+  // supported.
+  if (isa<arith::TruncIOp, arith::TruncFOp, arith::SelectOp>(op))
+    return false;
+
+  // Don't hoist through u1 -> fp casts as they aren't supported in
+  // ElementwiseOpToLLVM::reorderValues().
+  if (isa<arith::UIToFPOp>(op)) {
+    Type opType = getElementTypeOrSelf(op->getOperand(0));
+    if (opType.isInteger(1))
+      return false;
+  }
+
+  return true;
+}
+
+// Analog of canHoistDotOpEncV2, but for MMAv3 (WGMMA where operand A
+// is in registers).
+bool canHoistDotOpEncV3(Operation* op) {
```

Comment:
why is this any different than `canHoistDotOpEncV2`? I would expect it to be the same

Reply:
They're almost the same; the only difference is that MMAv3 hoisting doesn't support downcasting yet, because the lowering of shared-to-MMAv3-dotop-copy logic that I added doesn't yet support it.

For v3, I also added stricter checks like
```
  // Must have exactly one result and at least one operand
  if (op->getNumOperands() == 0 || op->getNumResults() != 1)
    return false;

  // Operands and results must be of RankedTensorType and Blocked or DotOp
  if (!(all_of(op->getOperandTypes(), isBlockedOrDotOpRankedTensor) &&
        all_of(op->getResultTypes(), isBlockedOrDotOpRankedTensor)))
    return false;
```

but left the v2 one intact in case something breaks.